doc/SAI-Proposal-Packet-Trimming.md

```diff
@@ -0,0 +1,143 @@
+     │              │                                                                                                                   
+     └──────────────┘                                                                                                                   
+```
+
+This feature assumes that forwarding tables are configured properly, and the original packet would be delivered to the destination successfully if not for the congestion.
+
+## Spec
+There is a tradeoff between trying to configure a higher threshold in a queue buffer profile and trimming the packet.
+
+If the user chooses to configure higher thresholds for queues, the probability of a drop on a particular queue is lower only if other ports are less congested at the moment.
+
+However, if all the ports are equally utilized, it makes sense to create a different buffer profile for these queues, with a stricter threshold to have more fairness in shared buffer.
+
+A static trimming threshold may not be effective with shared buffer switches, where the buffer resources allocated to a queue or port can vary over time. Therefore, we propose adding a new attribute to a buffer profile to allow configuring packet trimming on such stricter profiles:
+```
+/**
+ * @brief Enum defining queue actions in case of packet discard.
+ */
+typedef enum _sai_buffer_profile_packet_discard_action_t
+{
+    /**
+     * @brief Drop the packet.
+     *
+     * Default action. Packet has nowhere to go
+     * and will be dropped.
+     */
+    SAI_BUFFER_PROFILE_PACKET_DISCARD_ACTION_DROP = 0x00000000,
+
+    /**
+     * @brief Trim the packet.
+     *
+     * Try sending a shortened packet over a different
+     * queue.
+     */
+    SAI_BUFFER_PROFILE_PACKET_DISCARD_ACTION_TRIM = 0x00000001,
+} sai_buffer_profile_packet_discard_action_t;
+
+   /**
+     * @brief Buffer profile discard action
+     *
+     * Action to be taken upon packet discard due to
+     * buffer profile configuration. Applicable only
+     * when attached to a queue.
+     *
+     * @type sai_buffer_profile_packet_discard_action_t
+     * @flags CREATE_AND_SET
+     * @default SAI_BUFFER_PROFILE_PACKET_DISCARD_ACTION_DROP
+     */
+    SAI_BUFFER_PROFILE_ATTR_PACKET_DISCARD_ACTION,
+```
+
+Trimming engine attributes are configured globally.
+```
+   /**
+     * @brief Packet trimming size
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 128
+     */
+    SAI_SWITCH_ATTR_PACKET_TRIMMING_SIZE,
+
+    /**
+     * @brief New packet trimming DSCP value
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_PACKET_TRIMMING_DSCP_VALUE,
+
+    /**
+     * @brief New packet trimming queue index
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_PACKET_TRIMMING_QUEUE_INDEX,
```

Comment:
Typically the domain will be configured with TRIM_DSCP value and its much better to derive to the queue from the DSCP. Current proposal dis-associates the TRIM_DSCP and queue index i.e. one can configure DSCP_TRIM that may be handled by a certain queue in hw and SAI may configure wrong queue index.
Would it better to use qos map profile
SAI_QOS_MAP_TYPE_DSCP_TO_TC to SAI_QOS_MAP_TYPE_TC_TO_QUEUE.
Instead of hardcoding the queue index.

Reply:
Added two modes - mapping from QoS maps and straight setting with this attribute.